monkeytype/stubs.py

```diff
@@ -181,16 +181,21 @@ def get_imports_for_signature(sig: inspect.Signature) -> ImportMap:
     return imports
 
 
-def update_signature_args(sig: inspect.Signature, arg_types: Dict[str, type], has_self: bool) -> inspect.Signature:
+def update_signature_args(
+        sig: inspect.Signature,
+        arg_types: Dict[str, type],
+        has_self: bool,
+        ignore_existing_annotations: bool = False) -> inspect.Signature:
     """Update argument annotations with the supplied types"""
     params = []
     for arg_idx, name in enumerate(sig.parameters):
         param = sig.parameters[name]
         typ = arg_types.get(name)
-        # Don't touch pre-existing annotations and leave self un-annotated
-        if (typ is not None) and \
-           (param.annotation is inspect.Parameter.empty) and \
-           ((not has_self) or (arg_idx != 0)):
+        typ = inspect.Parameter.empty if (typ is None or typ is NoneType) else typ
```

Comment:
why did you feel the need to add the check for `NoneType` here? what behavior did you observe without that check that you thought was problematic?

IIUC a `None` here means we recorded no type for this parameter, whereas `NoneType` means we recorded a type of `None`. The former should result in no annotation, the latter should result in an annotation of `None`. So I don't think we should be checking for `NoneType` here.

Reply:
Previous, while I tried using `NoneType` in a test. You can test this on the master branch by modifying the `test_update_arg_with_anno` in `test_stubs.py` as follows.

```Python
    def test_update_arg_with_anno(self):
        """Leave existing arg annotations alone"""
        sig = Signature.from_callable(UpdateSignatureHelper.has_annos)
        sig = update_signature_args(sig, {'a': str, 'b': NoneType}, False)
        expected = Signature(
            parameters=[
                Parameter('a', Parameter.POSITIONAL_OR_KEYWORD, annotation=int),
                Parameter('b', Parameter.POSITIONAL_OR_KEYWORD, annotation=NoneType)
            ],
            return_annotation=int
        )
        assert sig == expected
```

I found that with the previous code, when a parameter has a trace suggesting `NoneType`, it will annotate it as `NoneType`. You will pass the test in this way. Thus, I thought that is because we only checked `None` but not `NoneType` in the `update_signature_arg` part, resulting in this behavior. I think as you mentioned in other part, the desire behavior is to return `None` in this way?

Reply:
Yes, the desired behavior when tracing gives `NoneType` is to annotate with `None`, but that's not what your change here does. Your change here causes that parameter to not be annotated at all.

The issue of `NoneType` vs `None` is separate from this enhancement and would best be handled in a separate PR for clarity. This line should only check `if typ is None` and remove the `NoneType` part.